### DISCUSSION TITLE
RFC: New Tanstack Integration

### DISCUSSION THREAD

**Discussion by juliusmarminge** [trpc/trpc#6240] [RFC / Ideas]
> 🔒 Conversation locked as this has been released - see the blog [post here](https://trpc.io/blog/introducing-tanstack-react-query-client) & leave thoughts in #6508 

As tRPC has grown, its types have grown with it. In particular, our React Query integration that we need to keep in sync with updates to the upstream React Query package whoses types are also quite complex with a lot of generics that need to match up. Further, complying to semver on the type level is extremely hard which can lead to breaking changes in patch or minor versions of either package.

## Our proposal

To address this, we've been exploring a more minimal integration package that greatly simplifies both runtime and type integration code between the two packages. This new package builds on top of the `queryOptions` function introduced not too long ago in React Query. Before we dive in to why, let's look at what will change:

### Before

```ts
import { createServerSideHelpers } from "@trpc/react-query/server"
import { trpc } from "~/utils/trpc"

export function loader()  {
  const trpc = createServerSideHelpers<AppRouter>()
  void trpc.myQuery.prefetch()
}

export function Component() {
  const utils = trpc.useUtils()
  const query = trpc.myQuery.useQuery()

  const mutation = trpc.myMutation.useMutation({
    onSettled: () => utils.myQuery.invalidate()
  })

  // ...
}
```

### After

```ts
import { useQuery, useMutation } from "@tanstack/react-query";
import { createTRPCOptionsProxy } from '@trpc/tanstack-react-query';
import { useTRPC } from "~/utils/trpc"

export function loader()  {
  const trpc = createTRPCOptionsProxy<AppRouter>()

  void trpc.queryClient.prefetchQuery(
    trpc.myQuery.queryOptions() // <--
  )
}

export function Component() {
  const trpc = useTRPC()

  const myQuery = trpc.myQuery.queryOptions()
  const query = useQuery(myQuery)

  const mutation = useMutation(
    trpc.myMutation.mutationOptions({
      onSettled: () => trpc.queryClient.invalidateQueries({
        queryKey: myQuery.queryKey
      })
    })
  )

  // ...
}
```

For reference, this is how raw React Query would look like in this example:

```ts
import { queryOptions, useQuery, useQueryClient, useMutation } from "@tanstack/react-query";
import { makeQueryClient } from "~/utils/queryClient"

const myQuery = queryOptions({
    queryKey: ["myQuery"],
    queryFn: () => fetchApi("...")
})

export function loader()  {
  const queryClient = makeQueryClient()
  void queryClient.prefetchQuery(myQuery)
}

export function Component() {
  const queryClient = useQueryClient();
  const query = useQuery(myQuery);

  const mutation = useMutation({
    mutationFn: () => { ... },
    onSettled: () => queryClient.invalidateQueries({
      queryKey: query.queryKey
    })
  })

  // ...
}
```

## Hold on, this seems like worse DX, no?

At first glance, you probably noticed that this new API is more verbose than previously. While this is true, it comes with a couple of advantages.

Firstly, the main limitation with out current package is that we do not support all of the queryClient methods. While [this list](https://trpc.io/docs/client/react/useUtils#helpers) has been improved over time, it generally takes user engagement for us to add support for a new method which can be frustrated for users who wants to use something _today_. Additionally, for each method we implement adds complexity to our already quite gnarly types which can easily touch internal types that might change in minor/patch React Query versions.

With this new package, we remove a lot of the complexity back to being _just React Query_. You have all their methods available without any abstraction hidden on top. This will also help onboarding new developers to your codebase, as you'll only have to teach them React Query. There'll be no tRPC layer on top. This also means we don't have to document or explain to people how React Query works, as you will use React Query directly.

Next, `eslint-plugin-react-hooks` have never been able to detect misuse of `trpc.proc.useQuery`, meaning you could call this hook conditionally without any static analysis tool catching it. tRPC also supports inverting the control of components (ie. polymorphism). This will not be possible under React Compiler unless we add these APIs. Calling hooks on props is entirely banned and won't compile with React Forget.

Additionally, we believe that with this approach, a new React Query major does not require a new tRPC major, and vice versa.

Lastly, this could potentially also make it trivial to add `@trpc/tanstack-vue-query` or similar in the future.

## Try it out today!

To make this easy for you to try out and provide your feedback, we added the `queryOptions` method to the existing proxy in `@trpc/react-query`:

```ts
export function Component() {
  const utils = trpc.useUtils()
  const query = useQuery(
    utils.myProc.queryOptions()
  )

  // ...
}
```

Note that this is a simple version and only includes `.queryOptions()`. You can checkout our [`Tanstack Start example`](https://github.com/trpc/trpc/tree/next/examples/tanstack-start) for how it might look to adopt this in your app, as well as [the PR](https://github.com/trpc/trpc/pull/6134) implementing the new package.

Please let us know your thoughts! We would hate to implement this and for no one to like it. Our plan is to deprecate the current `@trpc/react-query` package if all goes well, hopefully with codemod for easy migration! This API is not final and we're very much open to suggestions!

**Comment by juliusmarminge** [comment]
https://github.com/trpc/trpc/pull/6234#discussion_r1844070246

**Comment by Nick-Lucas** [comment]
The big question for me is where we draw lines on verbosity, React Query can be quite verbose on its own and typically wrapped inside named hooks when it's all used manually, but tRPC usage _should_ feel seamless and balance internal complexity with external complexity 

This is a good example of where we have tradeoffs to make:

<img width="420" alt="image" src="https://github.com/user-attachments/assets/f942f439-9364-4dcb-b29e-93b87b6548d7">

Option C is probably too far since it's wrapping React Query but is easily the nicest to use, option B uses a standard interface which can be used in a few QueryClient methods. Also needing to call both useTRPC and useQueryClient was very quickly an annoyance for me when [updating tests](https://github.com/trpc/trpc/pull/6234/files) and so I've played with exposing queryClient directly on the tRPC utils.

Bit of a balancing act to create a great DX while also developing a dogma which will guide future API surface area

**Reply by Nick-Lucas** [reply]
Maybe useTRPC's interface could be something like `const [trpc, queryClient] = useTRPC()` so that we don't end up needing the queryClient directly on the proxy but can simplify usage a bit?

**Reply by Nick-Lucas** [reply]
On reflection, if somebody does want a wrapper with both tRPC and queryClient in, they can make that in about 30 seconds

So maybe trpc can just have `queryKey()` and `queryFilter()` (dropping the get for consistency with `queryOptions`) on and the rest of the DX'ing can be trivially rolled

What do you think @juliusmarminge @KATT ?

**Reply by juliusmarminge** [reply]
Yea wrapper hook in user land seems alright to me. Maybe that means we should also drop nesting the queryClieny on the proxy altogether?

**Reply by TkDodo** [reply]
per design, we can pass `queryOptions` to invalidation, so:

```
queryClient.invalidateQueries(utils.myProc.queryOptions())
```

**Reply by Nick-Lucas** [reply]
Did not realise that! Albeit inputs would be mandatory when creating queryOptions and are not mandatory during invalidation 

So I think there's merit to having all 3 methods (queryKey, queryFilter, queryOptions) we've added since they give users options and represent interfaces which do exist in RQ

**Reply by TkDodo** [reply]
yeah I actually like that 👍

**Reply by juliusmarminge** [reply]
queryFilter could also take strongly typed options such as predicate functions etc

**Reply by Nick-Lucas** [reply]
Good point, I don't believe I played with that initially but we can definitely enhance it a bit more

**Comment by lunacrafts** [comment]
This might be relevant to this discussion: https://github.com/trpc/trpc/discussions/5751

At the moment, It's not possible to use browser-side tRPC & TanStack Query, that leads to significant disadvantages.

**Comment by Sliov** [comment]
is `prefetch` query the way to go over using the `ReactQueryStreamedHydration` [provider](https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr#experimental-streaming-without-prefetching-in-nextjs)?

**Reply by juliusmarminge** [reply]
There are some limitations in Next that makes the streamed provider not viable (mostly around accessing headers in the SSR phase) that the prefetching along with streaming promises solves

**Comment by alechaye** [comment]
I feel this is the way to go. This will provide more freedom and flexibility to the developers to implement complex use cases (for which raw react query is ideal). However, it would be nice to provide code snippets/helper functions in the docs that can be copied/pasted (like shadcn) to simplify the API for the most common use cases. That way, the library remains flexible and easy to evolve, and we could still avoid some verbosity.

**Reply by juliusmarminge** [reply]
Can you give some examples of what these snippets may look like? Not sure I understand

**Comment by rmarganti** [comment]
The current implementation works 90% of the time without thought. But that 10% comes with some big pain points (mostly revolving around query keys). I think exposing query options instead of a wrapper would be a big improvement.

**Reply by juliusmarminge** [reply]
Just to clarify you'd prefer if we keep the wrappers and have this as an extra thing?

**Reply by rmarganti** [reply]
No, I'd be okay with ditching the Tanstack Query wrapper all together. I much prefer flexibility and composability, even when it comes with a bit of verbosity.

**Reply by juliusmarminge** [reply]
alright good :)

**Reply by maxime-louellette** [reply]
I'm on your side with this. I was in discussion with my teams to entirely remove it because of these limitations. Even if it is really fun to use when it work, centralised hooks, while still verbose, have a really good dx and still goes really well with trpc

**Comment by Nsttt** [comment]
I see it as way nicer. Exposing the query keys right also gives the possibility of integrating new tooling around it. Very much in favor of the change.

**Comment by thomasballinger** [comment]
I think this is great! I like the queryOptions integration, that's how I plugged Convex into React Query in https://github.com/get-convex/convex-react-query.
```ts
const { isPending, error, data } = useQuery({
  ...convexQuery(api.repos.get, { repo: "made/up" }),
  gcTime: 10000, // unsubscribe after 10s of no use
});
```

**Reply by DogPawHat** [reply]
Yeah I loved what you did there.

**Reply by DogPawHat** [reply]
The main difference between convex and trpc seems to be the `queryClient` on the trpc proxy, whereas Convex just adds a default `queryFn` and `queryKeyHashFn`, and then has the `convexQuery` factory do the special sauce with making sure query and the convex sync stuff plays well together.

**Reply by juliusmarminge** [reply]
We probably wont have the queryClient on the proxy, both from feedback here but also from our own exploration. You can grab it from `useQueryClient` instead 

I don't know how convex works, but trpc client lives in state which is why you need the hook call to get the proxy

**Reply by DogPawHat** [reply]
I took a quick look at https://github.com/get-convex/convex-react-query/blob/main/src/index.ts to get an idea.

The magic is mostly in the `convexQueryClient.connect(tanstackQueryClient)` path, which take the convex api query key hashes, and set up a whole bunch of subscriptions to the normal WebSocket based `convexClient`  that calls `setQueryData` on updates. tRPC probably won't need that opinionated of a sync implementation I think.

**Comment by illispi** [comment]
Does this make porting/updating on third party frameworks like solidjs etc. easier for trpc/tantack query?

**Reply by Nick-Lucas** [reply]
In theory yes! We don't have a ton of experience with other frontend frameworks but since we're just interacting with interfaces native to TQ it would be much simpler than before.

Maybe we could release a framework agnostic "core" package which can be used to implement framework specific packages, similar to how Tanstack does things

**Reply by TkDodo** [reply]
One problem here is that `queryOptions` doesn’t live in the query core, but in the respective adapters, because the options are slightly different for vue/solid/react

**Reply by juliusmarminge** [reply]
Yes but given how much simpler this new package is compared to the old one we could potentially make those "copies" and still have a "maintainable amount"

Or a community member could take it on - the main thing is the reduced complexity make it much "friendlier" to tackle

**Comment by graup** [comment]
This is great and I can't wait to have it! This will make new features like [skip token](https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#typesafe-disabling-of-queries-using-skiptoken) available immediately without having to wait for TRPC to be updated. :+1:

One note, I think I might prefer writing "overrides" like this:

```ts
const mutation = useMutation({
  ...trpc.myMutation.mutationOptions(),
  onSettled: () => trpc.queryClient.invalidateQueries(trpc.myQuery.queryOptions()),
  onError: (error) => toast({ description: formatError(error) }),
})
```

(And I can easily do that because the API is very transparent.)

**Reply by Nick-Lucas** [reply]
Got to be a little careful with these overrides because we may occasionally hook into something internally, by passing the options in we can assign both without any undesired behaviour

I'd say having the _choice_ is a good thing though

**Reply by graup** [reply]
True. How about infiniteQueryOptions? From your tests I got this. Having two args is fine but not ideal IMO, especially for the "empty input" case.

```ts
const queryOptions = trpc.post.list.infiniteQueryOptions(
  {},
  {
    getNextPageParam(lastPage) {
      return lastPage.next;
    },
  },
);
const query1 = useInfiniteQuery(queryOptions);
```

Would this also work?

```ts
const queryOptions = trpc.post.list.infiniteQueryOptions();
const query1 = useInfiniteQuery({
  ...queryOptions,
  getNextPageParam(lastPage) {
    return lastPage.next;
  },
});
```

**Reply by juliusmarminge** [reply]
It should work, but you'd miss the added trpc options to the result which can be used eg for permission systems

**Comment by DogPawHat** [comment]
I had a pr open to add the `usePrefetchQuery` method to tRPC #5862 which required updating the TanStack query deps to the v5 minor that it had, which is a breaking change under, semver rules meant that if v11 wasn't in beta would have required a tRPC major (I know thing are changing a bit from v11 stable mind). It also meant there was like a 3-ish week gap between 'usePrefetchQuery' landing in Tanstack Query and me making the PR and then Julius reviewing it. just having the `queryOptions` api would have removed that hard coupling entirely. 

Given how nice the `queryOptions` API is working for Convex, I think it makes sense to use this route.

**Comment by infodusha** [comment]
A new API is promising
I find the benefit of hooks validation great as well as not mentioned thing:
It may allow to have a query that uses trpc mutation while still being type-safe (you need to remap options for this but it may work)

**Comment by shoooe** [comment]
I'm apparently the minority but I find the new syntax very unreadable. It feels like most of the lines are distracting implementation details. Even with those very simple examples I find myself reading so many words just to mentally parse that, e.g., I'm just mutating something and refetching a query.

Have you guys/gals explored the idea of maybe just exposing queryOptions and mutationOptions as a method on query/mutations so that we can keep the current DX but also allow for an escape hatch for things that aren't yet implemented?

Anyway, I totally understand if it's that burdensome to maintain a wrapper package.

While I'm at it, thanks to all contributors of this amazing library.

**Comment by heddendorp** [comment]
With tanStack query being extended to other platforms such as angular (https://tanstack.com/query/latest/docs/framework/angular/overview) it would be amazing to validate if this approach could be suitable for all platforms or if lightweight adjustments could be provided to make it work in all tanStack query environments.

**Comment by juliusmarminge** [comment]
https://trpc.io/blog/introducing-tanstack-react-query-client
